lib/internal/util.js

```diff
@@ -187,3 +191,42 @@ exports.convertToValidSignal = function convertToValidSignal(signal) {
 
   throw new Error('Unknown signal: ' + signal);
 };
+
+const kCustomPromisifiedSymbol = Symbol('util.promisify.custom');
+function promisify(orig) {
+  if (typeof orig !== 'function') {
+    throw new TypeError('The first argument to promisify() must be a function');
+  }
+
+  if (orig[kCustomPromisifiedSymbol]) {
+    return orig[kCustomPromisifiedSymbol];
+  }
+
+  if (isAsyncFunction(orig)) {
+    return orig;
+  }
+
+  const fn = function(...args) {
+    const promise = createPromise();
+    try {
+      orig.call(this, ...args, (err, data) => {
+        if (err) {
+          promiseReject(promise, err);
+        } else {
+          promiseResolve(promise, data);
+        }
+      });
+    } catch (err) {
+      promiseReject(promise, err);
+    }
+    return promise;
+  };
+
+  Object.setPrototypeOf(fn, Object.getPrototypeOf(orig));
```

Comment:
I'd make sure this doesn't decrease the performance of getting `fn`'s properties; if in doubt, an `if (Object.getPrototypeOf(fn) !== Object.getPrototypeOf(orig))` might be good.